homeassistant/components/mailbox/__init__.py

```diff
@@ -0,0 +1,256 @@
+        """Return the media blob for the msgid."""
+        raise NotImplementedError()
+
+    def get_messages(self):
+        """Return a list of the current messages."""
+        raise NotImplementedError()
+
+    def delete(self, msgids):
+        """Delete the specified messages."""
+        raise NotImplementedError()
+
+
+class StreamError(Exception):
+    """Media streaming exception."""
+
+    pass
+
+
+class MailboxView(HomeAssistantView):
+    """Base mailbox view."""
+
+    def __init__(self, mailboxes):
+        """Initialize a basic mailbox view."""
+        self.mailboxes = mailboxes
+
+    def get_mailbox(self, platform):
+        """Retrieve the specified mailbox."""
+        for mailbox in self.mailboxes:
+            if mailbox.name == platform:
+                return mailbox
+        return None
+
+
+class MailboxPlatformsView(MailboxView):
+    """View to return the list of mailbox platforms."""
+
+    url = "/api/mailbox/platforms"
+    name = "api:mailbox:platforms"
+
+    @asyncio.coroutine
+    def get(self, request):
+        """Retrieve list of platforms."""
+        platforms = []
+        for mailbox in self.mailboxes:
+            platforms.append(mailbox.name)
+        return self.json(platforms)
+
+
+class MailboxMessageView(MailboxView):
+    """View to return the list of messages."""
+
+    url = "/api/mailbox/messages/{platform}"
+    name = "api:mailbox:messages"
+
+    @asyncio.coroutine
+    def get(self, request, platform):
+        """Retrieve messages."""
+        mailbox = self.get_mailbox(platform)
+        if mailbox is None:
+            return web.Response(status=401)
+        return self.json(mailbox.get_messages())
+
+
+class MailboxDeleteView(MailboxView):
+    """View to delete selected messages."""
+
+    url = "/api/mailbox/delete/{platform}"
+    name = "api:mailbox:delete"
+
+    @asyncio.coroutine
+    def post(self, request, platform):
+        """Delete items."""
+        mailbox = self.get_mailbox(platform)
+        if mailbox is None:
+            return web.Response(status=401)
+        try:
+            data = yield from request.json()
+            mailbox.delete(data)
+        except ValueError:
+            return self.json_message('Bad item id', HTTP_BAD_REQUEST)
+
+
+class MailboxMediaView(MailboxView):
+    """View to return a media file."""
+
+    url = r"/api/mailbox/media/{platform}/{msgid}"
+    name = "api:asteriskmbox:media"
+
+    @asyncio.coroutine
+    def get(self, request, platform, msgid):
+        """Retrieve media."""
+        mailbox = self.get_mailbox(platform)
+        if mailbox is None:
+            return web.Response(status=401)
+
+        hass = request.app['hass']
+        with suppress(asyncio.CancelledError, asyncio.TimeoutError):
+            with async_timeout.timeout(10, loop=hass.loop):
+                try:
+                    stream = yield from hass.async_add_job(
+                        partial(mailbox.get_media, msgid))
```

Comment:
Don't use partial, it messes with the annotations and ends up running everything in the executor. async_add_job already supports passing arguments as extra arguments:

```python
yield from hass.async_add_job(mailbox.get_media, msgid)
```